src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -705,12 +705,24 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             .iter()
             .filter_map(|expr| tables.node_type_opt(expr.hir_id))
             .map(|ty| self.resolve_vars_if_possible(&ty));
-        let (last_ty, all_returns_have_same_type) = ret_types.clone().fold(
-            (None, true),
-            |(last_ty, mut same): (std::option::Option<Ty<'_>>, bool), ty| {
+        let (last_ty, all_returns_have_same_type, count) = ret_types.clone().fold(
+            (None, true, 0),
+            |(last_ty, mut same, count): (std::option::Option<Ty<'_>>, bool, usize), ty| {
                 let ty = self.resolve_vars_if_possible(&ty);
-                same &= last_ty.map_or(true, |last_ty| last_ty == ty) && ty.kind != ty::Error;
-                (Some(ty), same)
+                same &= last_ty.map_or(true, |last_ty| {
+                    // FIXME: ideally we would use `can_coerce` here instead, but `typeck` comes
+                    // *after* in the dependency graph.
+                    match (&ty.kind, &last_ty.kind) {
```

Comment:
You could compare `mem::discriminant`s instead.

Reply:
By writing this match like this I am ignoring other types of inferred types and only considering equals float and integer literals. Using `mem::discriminant(ty.kind) == mem::discriminant(last_ty.kind)` only would cause two different array types to be considered equals, which we wouldn't want. Am I missing something?

Reply:
Can't you use `(Infer(a), Infer(b)) => discriminant(a) == discriminant(b)`?

Reply:
`InferTy` also has `TyVar` which I would want to exclude:
```
#[derive(HashStable)]
pub enum InferTy {
    TyVar(TyVid),
    IntVar(IntVid),
    FloatVar(FloatVid),

    /// A `FreshTy` is one that is generated as a replacement for an
    /// unbound type variable. This is convenient for caching etc. See
    /// `infer::freshen` for more details.
    FreshTy(u32),
    FreshIntTy(u32),
    FreshFloatTy(u32),
}
```

I'll try one thing and you can tell me if you prefer it.

Reply:
Check the last commit and tell me what you think.

Reply:
Seems like an improvement 👍